[GPU Process] [FormControls] Route drawing the ControlPart through GraphicsContext

shallawa opened the pull request:
#### 27e83068fe9e1feb64ac96d8f8b0c76d43a5e0fc
<pre>
[GPU Process] [FormControls] Route drawing the ControlPart through GraphicsContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=249248">https://bugs.webkit.org/show_bug.cgi?id=249248</a>
rdar://103315944

Reviewed by Aditya Keerthi and Simon Fraser.

We will make RenderTheme create a ControlPart from DOM element of RenderBox. We
will start with MeterPart as an example for the WebProcess drawing workflow.

RenderTheme::paint() will call GraphicsContext::drawControlPart() which will call
ControlPart::draw(). ControlPart::draw() will create the PlatformControl using its
ControlFactory. On macOS the ControlFactory will be of type ControlFactoryMac and
the PlatformControl will be of type MeterMac. ControlPart::draw() will call
MeterMac::draw() which will call AppKit to draw the NSLevelIndicatorCell.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawControlPart):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/controls/ControlFactory.cpp: Added.
(WebCore::ControlFactory::createControlFactory):
(WebCore::ControlFactory::sharedControlFactory):
* Source/WebCore/platform/graphics/controls/ControlFactory.h: Added.
* Source/WebCore/platform/graphics/controls/ControlPart.cpp: Added.
(WebCore::ControlPart::ControlPart):
(WebCore::ControlPart::controlFactory const):
(WebCore::ControlPart::platformControl const):
(WebCore::ControlPart::sizeForBounds):
(WebCore::ControlPart::draw const):
* Source/WebCore/platform/graphics/controls/ControlPart.h: Added.
(WebCore::ControlPart::type const):
(WebCore::ControlPart::setControlFactory):
* Source/WebCore/platform/graphics/controls/ControlPartType.cpp: Renamed from Source/WebCore/platform/graphics/formcontrols/ControlPartType.cpp.
* Source/WebCore/platform/graphics/controls/ControlPartType.h: Renamed from Source/WebCore/platform/graphics/formcontrols/ControlPartType.h.
* Source/WebCore/platform/graphics/controls/ControlStyle.h: Added.
* Source/WebCore/platform/graphics/controls/MeterPart.cpp: Added.
(WebCore::MeterPart::create):
(WebCore::MeterPart::MeterPart):
(WebCore::MeterPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/MeterPart.h: Added.
(WebCore::MeterPart::gaugeRegion const):
(WebCore::MeterPart::value const):
(WebCore::MeterPart::minimum const):
(WebCore::MeterPart::maximum const):
* Source/WebCore/platform/graphics/controls/PlatformControl.h: Added.
(WebCore::PlatformControl::PlatformControl):
(WebCore::PlatformControl::setFocusRingClipRect):
(WebCore::PlatformControl::updateCellStates):
(WebCore::PlatformControl::sizeForBounds const):
(WebCore::PlatformControl::draw):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm: Added.
(WebCore::ControlFactory::createControlFactory):
(WebCore::ControlFactoryMac::drawingView const):
(WebCore::ControlFactoryMac::levelIndicatorCell const):
(WebCore::ControlFactoryMac::createPlatformMeter):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm: Added.
(WebCore::ControlMac::ControlMac):
(WebCore::ControlMac::setFocusRingClipRect):
(WebCore::ControlMac::updateCellStates):
(WebCore::drawFocusRing):
(WebCore::drawCellOrFocusRing):
(WebCore::ControlMac::draw):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h: Added.
(WebCore::MeterMac::owningMeterPart const):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm: Added.
(WebCore::MeterMac::MeterMac):
(WebCore::MeterMac::updateCellStates):
(WebCore::MeterMac::sizeForBounds const):
(WebCore::MeterMac::draw):
* Source/WebCore/platform/graphics/mac/controls/WebControlView.h: Added.
* Source/WebCore/platform/graphics/mac/controls/WebControlView.mm: Added.
(+[WebControlWindow hasKeyAppearance]):
(+[WebControlWindow setHasKeyAppearance:]):
(-[WebControlWindow hasKeyAppearance]):
(-[WebControlWindow isKeyWindow]):
(-[WebControlWindow _needsToResetDragMargins]):
(-[WebControlWindow _setNeedsToResetDragMargins:]):
(+[WebControlView clipBounds]):
(+[WebControlView setClipBounds:]):
(-[WebControlView init]):
(-[WebControlView window]):
(-[WebControlView isFlipped]):
(-[WebControlView currentEditor]):
(-[WebControlView _automaticFocusRingDisabled]):
(-[WebControlView _focusRingVisibleRect]):
(-[WebControlView _focusRingClipAncestor]):
(-[WebControlView addSubview:]):
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::ensuredView):
(WebCore::ThemeMac::useFormSemanticContext):
(WebCore::ThemeMac::setUseFormSemanticContext):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::ensureControlPart):
(WebCore::RenderBox::paintBoxDecorations):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderMeter.cpp:
(WebCore::RenderMeter::updateLogicalWidth):
(WebCore::RenderMeter::computeLogicalHeight const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createMeterPartForRenderer):
(WebCore::RenderTheme::createControlPartForRenderer const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::paint):
(WebCore::RenderTheme::meterSizeForBounds const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::canCreateControlPartForRenderer const):
(WebCore::RenderTheme::useFormSemanticContext const):
(WebCore::RenderTheme::supportsLargeFormControls const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::useFormSemanticContext const):
(WebCore::RenderThemeMac::supportsLargeFormControls const):
(WebCore::RenderThemeMac::meterSizeForBounds const):
(WebCore::RenderThemeMac::paintMeter): Deleted.
(WebCore::RenderThemeMac::levelIndicatorFor const): Deleted.
* Source/WebCore/rendering/RenderThemeWin.cpp:
(WebCore::RenderThemeWin::meterSizeForBounds const): Deleted.
* Source/WebCore/rendering/RenderThemeWin.h:

Canonical link: <a href="https://commits.webkit.org/257981@main">https://commits.webkit.org/257981@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e09b28ac4d7068459c6ac1e4af9cac8f2db47b50

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9682 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33582 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109845 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170135 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104526 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10616 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92935 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107700 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8031 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91276 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34639 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/89943 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22665 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3406 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24183 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3414 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9527 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43674 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5470 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5212 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/controls/ControlStyle.h:
```diff
@@ -0,0 +1,84 @@
+template<class Encoder>
+void ControlStyle::encode(Encoder& encoder) const
+{
+    encoder << fontSize;
```
This should probably use `WebCoreArgumentCoders.serialization.in`, which the the new way of writing encoders/decoders.

Same comment about other serialization in this patch.

shallawa replied:
I removed all the encoding and decoding from this patch since they are not used. I will add them in the next patch where drawing the ControlPart in GPUP is enabled.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
```diff
@@ -0,0 +1,69 @@
+    // Use a fake view.
+    [m_drawingView setFrameSize:NSSizeFromCGSize(size)];
+    ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    [m_drawingView setAppearance:[NSAppearance currentAppearance]];
```
There's another (recently added) [bit](https://github.com/WebKit/WebKit/blob/e48228395c2e9e0b09f86503efa1231cec055d09/Source/WebCore/platform/mac/ThemeMac.mm#L703) we need to propagate to the drawing view.

shallawa replied:
Done. I added OptionSet<Flags> to ControlStyle. The flag ControlStyle::Flag::SemanticContext is used for this purpose.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/WebFormControlView.mm:
```diff
@@ -0,0 +1,151 @@
+
+@end
+
+@implementation WebCoreTextFieldCell1
```
Can we share this code with `RenderThemeMac` – maybe that file could import this `WebFormControlView.h`?

I worry we will forget to rename `WebCoreTextFieldCell1`.

shallawa replied:
I removed this interface from WebControlView.h since it is not used in this patch. It will be added when TextField Control is added.

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1605,10 +1606,14 @@ void RenderBox::paintBoxDecorations(PaintInfo& paintInfo, const LayoutPoint& pai
     if (style().hasEffectiveAppearance()) {
-        ControlStates* controlStates = controlStatesForRenderer(*this);
-        borderOrBackgroundPaintingIsNeeded = theme().paint(*this, *controlStates, paintInfo, paintRect);
-        if (controlStates->needsRepaint())
```
Will there be a replacement for this with the new approach?

It's possible this is not even necessary with the old implementation, but we would need to retest focus ring behavior.

shallawa replied:
Meter ControlPart doesn't set needsRepaint. So this change is safe for now. I will test the focus ring behavior for the upcoming ControlParts and add similar behavior if it is needed.

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1605,10 +1606,14 @@ void RenderBox::paintBoxDecorations(PaintInfo& paintInfo, const LayoutPoint& pai
-        if (controlStates->needsRepaint())
-            view().scheduleLazyRepaint(*this);
+        if (auto* control = theme().controlPartForRenderer(*this))
+            borderOrBackgroundPaintingIsNeeded = theme().paint(*this, *control, paintInfo, paintRect);
```
Notes for subsequent patches:
- You might need to put `controlStates` somewhere here.
- There's also `theme().paintDecorations()` a couple of lines below, that will need to be converted the new architecture.

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1605,10 +1606,14 @@ void RenderBox::paintBoxDecorations(PaintInfo& paintInfo, const LayoutPoint& pai
-        borderOrBackgroundPaintingIsNeeded = theme().paint(*this, *controlStates, paintInfo, paintRect);
-        if (controlStates->needsRepaint())
-            view().scheduleLazyRepaint(*this);
+        if (auto* control = theme().controlPartForRenderer(*this))
```
I was going to ask why we can't put everything in the new `paint`, but then it would be tricker to distinguish whether we're going down the new path and also set `borderOrBackgroundPaintingIsNeeded`. I think this is fine.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -456,6 +457,140 @@ ControlPartType RenderTheme::autoAppearanceForElement(RenderStyle& style, const
+    case ControlPartType::MenulistButton:
+        break;
+
+    case ControlPartType::Meter: {
```
We may want to split this up into separate methods like the old `paint` – I worry this method will become very large.

shallawa replied:
Done. A separate function is added to create the MeterPart. I can also use EnumeratedArray if it makes things clearer.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -456,6 +457,140 @@ ControlPartType RenderTheme::autoAppearanceForElement(RenderStyle& style, const
+        break;
+
+    case ControlPartType::Meter: {
+        const auto& renderMeter = downcast<RenderMeter>(renderer);
```
`ASSERT(is<RenderMeter>(renderer))`?

This wasn't always the case before we implemented the new `appearance` spec, but I think it's safe now due to `RenderTheme::adjustAppearanceForElement`.

shallawa replied:
The ASSERT was added.

smfr commented on Source/WebCore/platform/graphics/GraphicsContext.h:
```diff
@@ -269,6 +270,8 @@ class GraphicsContext {
     virtual void drawPattern(NativeImage&, const FloatRect& destRect, const FloatRect& tileRect, const AffineTransform& patternTransform, const FloatPoint& phase, const FloatSize& spacing, const ImagePaintingOptions& = { }) = 0;
     WEBCORE_EXPORT virtual void drawPattern(ImageBuffer&, const FloatRect& destRect, const FloatRect& tileRect, const AffineTransform& patternTransform, const FloatPoint& phase, const FloatSize& spacing, const ImagePaintingOptions& = { });
 
+    WEBCORE_EXPORT virtual void drawControlPart(ControlPart&, const FloatRect&, float deviceScaleFactor, const ControlStyle&);
```
`const ControlPart&`

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/controls/ControlStyle.h:
```diff
@@ -0,0 +1,55 @@
+        Indeterminate   = 1 << 7,
+        SpinUp          = 1 << 8, // Sub-state for HoverState and PressedState.
+        Presenting      = 1 << 9,
+        SemanticContext = 1 << 10,
```
`FormSemanticContext`?

Couple of other places that are `useSemanticContext` should probably be `useFormSemanticContext` too.

shallawa replied:
Fixed. FormSemanticContext and useFormSemanticContext will be used.

smfr commented on Source/WebCore/platform/graphics/controls/ControlPart.h:
```diff
@@ -0,0 +1,63 @@
+    PlatformControl* platformControl();
+    virtual std::unique_ptr<PlatformControl> createPlatformControl() = 0;
+
+    ControlPartType m_type;
```
This can be const

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/controls/ControlPart.h:
```diff
@@ -0,0 +1,63 @@
+
+    ControlPartType type() const { return m_type; }
+
+    ControlFactory& controlFactory() const;
```
Will there ever be more than one factory? The factory is a singleton I think so why does this class need a getter/setter for the factory?

shallawa replied:
We need it for GPUP. In GPUP ControlParts can be rendered simultaneously in multiple threads. So we need every thread to have its own factory. Every RemoteDisplayListRecorder will set its factory to the decoded ControlPart before displaying it.

smfr commented on Source/WebCore/platform/graphics/controls/ControlStyle.h:
```diff
@@ -0,0 +1,55 @@
+namespace WebCore {
+
+struct ControlStyle {
+    enum class Flag : uint16_t {
```
`State` might be a better name. Or `Attribute`.

shallawa replied:
Fixed. `Flag` is renamed to `State`.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
```diff
@@ -0,0 +1,74 @@
+    [m_drawingView setAppearance:[NSAppearance currentAppearance]];
+    ALLOW_DEPRECATED_DECLARATIONS_END
+#if USE(NSVIEW_SEMANTICCONTEXT)
+    if ([WebControlView useSemanticContext])
```
Seems a little strange that this is read off a class method, can we pass `ControlStyle` as a parameter into this method instead?

shallawa replied:
Fixed. ControlStyle is used instead.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -0,0 +1,122 @@
+void ControlMac::updateCellStates(const FloatRect&, const ControlStyle& style)
+{
+    [WebControlWindow setHasKeyAppearance:!style.flags.contains(ControlStyle::Flag::WindowInactive)];
+    [WebControlView setUseSemanticContext:style.flags.contains(ControlStyle::Flag::WindowInactive)];
```
Incorrect flag used here, but this line should be irrelevant if you make the other changes described.

shallawa replied:
This line is removed and `[WebControlView setUseSemanticContext]` is also removed.

smfr commented on Source/WebCore/platform/graphics/controls/PlatformControl.h:
```diff
@@ -0,0 +1,60 @@
+    virtual void draw(GraphicsContext&, const FloatRect&, float, const ControlStyle&) { }
+
+protected:
+    ControlPart& m_part;
```
To clarify ownership, I'd call this `m_owningPart`.

shallawa replied:
Fixed. The member is renamed to `m_owningPart`.

smfr commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
```diff
@@ -0,0 +1,74 @@
+NSLevelIndicatorCell* ControlFactoryMac::levelIndicatorCell() const
+{
+    if (!m_levelIndicatorCell) {
+        m_levelIndicatorCell = adoptNS([[NSLevelIndicatorCell alloc] initWithLevelIndicatorStyle:NSLevelIndicatorStyleContinuousCapacity]);
```
Why is this long-lived, and not just owned by a PlatformControl?

shallawa replied:
I do not know but I copied this from the old behavior. It might be expensive to create an NSCell for every renderer. I will try later to remove this shared NSCell and see the performance.

smfr commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
```diff
@@ -0,0 +1,57 @@
+    void updateCellStates(const FloatRect&, const ControlStyle&) override;
+
+    void drawCellOrFocusRing(GraphicsContext&, const FloatRect&, const ControlStyle&, NSCell *, NSView *, bool drawCell = true);
+    void drawCellOrFocusRing(GraphicsContext&, const FloatRect&, float deviceScaleFactor, const ControlStyle&, NSCell *, NSView *, bool drawCell = true);
```
These two functions have surprisingly different behavior; one can use an intermediate buffer. The other one will just draw a wrongly scaled thing in scaled contexts. Either give them different names, or make one of them private.

shallawa replied:
Fixed. I renamed the one that creates the ImageBuffer to `ControlMac::draw()`. I kept the second one as is but I removed it as a class method and I made it a static function.

smfr commented on Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
```diff
@@ -0,0 +1,106 @@
+
+    drawCellOrFocusRing(context, rect, deviceScaleFactor, style, m_levelIndicatorCell.get(), view);
+
+    [m_levelIndicatorCell setControlView:nil];
```
Where is the corresponding call to `[m_levelIndicatorCell setControlView:]`? Can we put it in this function too, for symmetry?

pxlcoder replied:
It happens implicitly in `[cell drawWithFrame:rect inView:view];`.

shallawa replied:
I think the call that sets the `NSView` to the `NSCell` is in `[NSCell drawWithFrame]`. I searched the WebCore source code and I can only see that we call `setControlView` with `nil` only. So I moved to drawCellOrFocusRing in ControlMac.mm

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -176,6 +176,7 @@ void RenderBox::willBeDestroyed()
 
     view().unscheduleLazyRepaint(*this);
     removeControlStatesForRenderer(*this);
+    theme().removeControlPartForRenderer(*this);
```
Shame that we place this cleanup burden on RenderBox. Can we do this some other way?

shallawa replied:
Fixe. This line was removed since the renderer `ControlPart` was moved to `RenderObjectRareData`.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -456,6 +457,189 @@ ControlPartType RenderTheme::autoAppearanceForElement(RenderStyle& style, const
+        break;
+
+    case ControlPartType::Meter:
+        return creatMeterPartForRenderer(renderer);
```
Typo: `creat`.

shallawa replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -456,6 +457,189 @@ ControlPartType RenderTheme::autoAppearanceForElement(RenderStyle& style, const
+
+static auto& controlPartRendererMap()
+{
+    using ControlPartRendererMap = HashMap<const RenderObject*, Ref<ControlPart>>;
```
Should we store references to ControlParts in `RenderObjectRareData` instead?

shallawa replied:
Done. A new data member of type `RefPtr<ControlPart>` was added to `RenderObjectRareData`.

smfr commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -456,6 +457,189 @@ ControlPartType RenderTheme::autoAppearanceForElement(RenderStyle& style, const
+    controlPartRendererMap().remove(&renderer);
+}
+
+OptionSet<ControlStyle::Flag> RenderTheme::extractControlStyleFlagsForRenderer(const RenderObject& o) const
```
`o` -> `renderer`

shallawa replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -456,6 +457,189 @@ ControlPartType RenderTheme::autoAppearanceForElement(RenderStyle& style, const
+    };
+}
+
+bool RenderTheme::paint(const RenderBox& box, ControlPart& part, const PaintInfo& paintInfo, const LayoutRect& rect)
```
`const ControlPart&` if possible

shallawa replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderTheme.h:
```diff
@@ -69,10 +69,20 @@ class RenderTheme {
-    // text of a button, is always rendered by the engine itself.  The boolean return value indicates
+    virtual bool canCreateControlPartForRenderer(const RenderObject&) const { return false; }
+    RefPtr<ControlPart> createControlPartForRenderer(const RenderObject&) const;
+    ControlPart* controlPartForRenderer(const RenderBox&) const;
```
is this `existingControlPartForRenderer` or `ensureControlPartForRenderer`?

shallawa replied:
It is `ensureControlPartForRenderer` but it will return `ControlPart*` because only `Meter` is capable of creating its `MeterPart` on macOS right now.

shallawa committed:
[GPU Process] [FormControls] Route drawing the ControlPart through GraphicsContext
https://bugs.webkit.org/show_bug.cgi?id=249248
rdar://103315944

Reviewed by Aditya Keerthi and Simon Fraser.

We will make RenderTheme create a ControlPart from DOM element of RenderBox. We
will start with MeterPart as an example for the WebProcess drawing workflow.

RenderTheme::paint() will call GraphicsContext::drawControlPart() which will call
ControlPart::draw(). ControlPart::draw() will create the PlatformControl using its
ControlFactory. On macOS the ControlFactory will be of type ControlFactoryMac and
the PlatformControl will be of type MeterMac. ControlPart::draw() will call
MeterMac::draw() which will call AppKit to draw the NSLevelIndicatorCell.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawControlPart):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/controls/ControlFactory.cpp: Added.
(WebCore::ControlFactory::createControlFactory):
(WebCore::ControlFactory::sharedControlFactory):
* Source/WebCore/platform/graphics/controls/ControlFactory.h: Added.
* Source/WebCore/platform/graphics/controls/ControlPart.cpp: Added.
(WebCore::ControlPart::ControlPart):
(WebCore::ControlPart::controlFactory const):
(WebCore::ControlPart::platformControl const):
(WebCore::ControlPart::sizeForBounds):
(WebCore::ControlPart::draw const):
* Source/WebCore/platform/graphics/controls/ControlPart.h: Added.
(WebCore::ControlPart::type const):
(WebCore::ControlPart::setControlFactory):
* Source/WebCore/platform/graphics/controls/ControlPartType.cpp: Renamed from Source/WebCore/platform/graphics/formcontrols/ControlPartType.cpp.
* Source/WebCore/platform/graphics/controls/ControlPartType.h: Renamed from Source/WebCore/platform/graphics/formcontrols/ControlPartType.h.
* Source/WebCore/platform/graphics/controls/ControlStyle.h: Added.
* Source/WebCore/platform/graphics/controls/MeterPart.cpp: Added.
(WebCore::MeterPart::create):
(WebCore::MeterPart::MeterPart):
(WebCore::MeterPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/MeterPart.h: Added.
(WebCore::MeterPart::gaugeRegion const):
(WebCore::MeterPart::value const):
(WebCore::MeterPart::minimum const):
(WebCore::MeterPart::maximum const):
* Source/WebCore/platform/graphics/controls/PlatformControl.h: Added.
(WebCore::PlatformControl::PlatformControl):
(WebCore::PlatformControl::setFocusRingClipRect):
(WebCore::PlatformControl::updateCellStates):
(WebCore::PlatformControl::sizeForBounds const):
(WebCore::PlatformControl::draw):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm: Added.
(WebCore::ControlFactory::createControlFactory):
(WebCore::ControlFactoryMac::drawingView const):
(WebCore::ControlFactoryMac::levelIndicatorCell const):
(WebCore::ControlFactoryMac::createPlatformMeter):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h: Added.
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm: Added.
(WebCore::ControlMac::ControlMac):
(WebCore::ControlMac::setFocusRingClipRect):
(WebCore::ControlMac::updateCellStates):
(WebCore::drawFocusRing):
(WebCore::drawCellOrFocusRing):
(WebCore::ControlMac::draw):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h: Added.
(WebCore::MeterMac::owningMeterPart const):
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm: Added.
(WebCore::MeterMac::MeterMac):
(WebCore::MeterMac::updateCellStates):
(WebCore::MeterMac::sizeForBounds const):
(WebCore::MeterMac::draw):
* Source/WebCore/platform/graphics/mac/controls/WebControlView.h: Added.
* Source/WebCore/platform/graphics/mac/controls/WebControlView.mm: Added.
(+[WebControlWindow hasKeyAppearance]):
(+[WebControlWindow setHasKeyAppearance:]):
(-[WebControlWindow hasKeyAppearance]):
(-[WebControlWindow isKeyWindow]):
(-[WebControlWindow _needsToResetDragMargins]):
(-[WebControlWindow _setNeedsToResetDragMargins:]):
(+[WebControlView clipBounds]):
(+[WebControlView setClipBounds:]):
(-[WebControlView init]):
(-[WebControlView window]):
(-[WebControlView isFlipped]):
(-[WebControlView currentEditor]):
(-[WebControlView _automaticFocusRingDisabled]):
(-[WebControlView _focusRingVisibleRect]):
(-[WebControlView _focusRingClipAncestor]):
(-[WebControlView addSubview:]):
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::ensuredView):
(WebCore::ThemeMac::useFormSemanticContext):
(WebCore::ThemeMac::setUseFormSemanticContext):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::ensureControlPart):
(WebCore::RenderBox::paintBoxDecorations):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderMeter.cpp:
(WebCore::RenderMeter::updateLogicalWidth):
(WebCore::RenderMeter::computeLogicalHeight const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createMeterPartForRenderer):
(WebCore::RenderTheme::createControlPartForRenderer const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::paint):
(WebCore::RenderTheme::meterSizeForBounds const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::canCreateControlPartForRenderer const):
(WebCore::RenderTheme::useFormSemanticContext const):
(WebCore::RenderTheme::supportsLargeFormControls const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::useFormSemanticContext const):
(WebCore::RenderThemeMac::supportsLargeFormControls const):
(WebCore::RenderThemeMac::meterSizeForBounds const):
(WebCore::RenderThemeMac::paintMeter): Deleted.
(WebCore::RenderThemeMac::levelIndicatorFor const): Deleted.
* Source/WebCore/rendering/RenderThemeWin.cpp:
(WebCore::RenderThemeWin::meterSizeForBounds const): Deleted.
* Source/WebCore/rendering/RenderThemeWin.h:

Canonical link: https://commits.webkit.org/257981@main
